File: ProtoChat/protobuf3/ProtoChatWorker.cs
using System;
using System.Threading;
using System.Net.Sockets;

using chat.services;
using Chat.Protocol;
using Google.Protobuf;

namespace protobuf
{
   public class ProtoV3ChatWorker:IChatObserver
    {
        private IChatServer server;
		private TcpClient connection;

		private NetworkStream stream;
		private volatile bool connected;
		public ProtoV3ChatWorker(IChatServer server, TcpClient connection)
		{
			this.server = server;
			this.connection = connection;
			try
			{
				
				stream=connection.GetStream();
				connected=true;
			}
			catch (Exception e)
			{
                Console.WriteLine(e.StackTrace);
			}
		}

		public virtual void run()
		{
			while(connected)
			{
				try
				{
					
                    ChatRequest request = ChatRequest.Parser.ParseDelimitedFrom(stream);
					ChatResponse response =handleRequest(request);
					if (response!=null)
					{
					   sendResponse(response);
					}
				}
				catch (Exception e)
				{
                    Console.WriteLine(e.StackTrace);
				}
				
				try
				{
					Thread.Sleep(1000);
				}
				catch (Exception e)
				{
                    Console.WriteLine(e.StackTrace);
				}
			}
			try
			{
				stream.Close();
				connection.Close();
			}
			catch (Exception e)
			{
				Console.WriteLine("Error "+e);
			}
		}
		public virtual void messageReceived(chat.model.Message message)
		{
			Console.WriteLine("Message received  "+message);
			try
			{
				sendResponse(ProtoUtils.createNewMessageResponse(message));
			}
			catch (Exception e)
			{
				throw new ChatException("Sending error: "+e);
			}
		}

	public virtual void friendLoggedIn(chat.model.User friend)
		{
			Console.WriteLine("Friend logged in "+friend);
			try
			{
				sendResponse(ProtoUtils.createFriendLoggedInResponse(friend));
			}
			catch (Exception e)
			{
                Console.WriteLine(e.StackTrace);
			}
		}
		public virtual void friendLoggedOut(chat.model.User friend)
		{
			Console.WriteLine("Friend logged out "+friend);
			try
			{
				sendResponse(ProtoUtils.createFriendLoggedOutResponse(friend));
			}
			catch (Exception e)
			{
                Console.WriteLine(e.StackTrace);
			}
		}

		private ChatResponse handleRequest(ChatRequest request)
		{
			ChatResponse response =null;
            ChatRequest.Types.Type reqType=request.Type;
            switch(reqType){
                case ChatRequest.Types.Type.Login:
			{
				Console.WriteLine("Login request ...");
				chat.model.User user =ProtoUtils.getUser(request);
				try
                {
                    lock (server)
                    {
                        server.login(user, this);
                    }
					return ProtoUtils.createOkResponse();
				}
				catch (ChatException e)
				{
					connected=false;
					return ProtoUtils.createErrorResponse(e.Message);
				}
			}
                case ChatRequest.Types.Type.Logout:
			{
				Console.WriteLine("Logout request");
                chat.model.User user = ProtoUtils.getUser(request);
				try
				{
                    lock (server)
                    {

                        server.logout(user, this);
                    }
					connected=false;
					return ProtoUtils.createOkResponse();

				}
				catch (ChatException e)
				{
				   return ProtoUtils.createErrorResponse(e.Message);
				}
			}
                case ChatRequest.Types.Type.SendMessage:
			{
				Console.WriteLine("SendMessageRequest ...");
                chat.model.Message message = ProtoUtils.getMessage(request);
				try
				{
                    lock (server)
                    {
                        server.sendMessage(message);
                    }
                        return ProtoUtils.createOkResponse();
				}
				catch (ChatException e)
				{
					return ProtoUtils.createErrorResponse(e.Message);
				}
			}

                case ChatRequest.Types.Type.GetLoggedFriends:
			{
				Console.WriteLine("GetLoggedFriends Request ...");
                chat.model.User user = ProtoUtils.getUser(request);  //DTOUtils.getFromDTO(udto);
				try
				{
                    chat.model.User[] friends;
                    lock (server)
                    {

                         friends = server.getLoggedFriends(user);
                    }
					return ProtoUtils.createLoggedFriendsResponse(friends);
				}
				catch (ChatException e)
				{
					return ProtoUtils.createErrorResponse(e.Message);
				}
			}
            }
			return response;
		}

	private void sendResponse(ChatResponse response)
		{
			Console.WriteLine("sending response "+response);
			lock (stream)
			{
				response.WriteDelimitedTo(stream);
				stream.Flush();
			}

		}
	}

        
    
}

File: ChatProtobuff_v3/ChatNetworkingProtobuff/src/main/java/chat/network/protobuffprotocol/ProtoChatWorker.java
package chat.network.protobuffprotocol;

import chat.model.Message;
import chat.model.User;

import chat.services.ChatException;
import chat.services.IChatObserver;
import chat.services.IChatServer;

import java.io.*;
import java.net.Socket;


public class ProtoChatWorker implements Runnable, IChatObserver {
    private IChatServer server;
     private Socket connection;

     private InputStream input;
     private OutputStream output;
     private volatile boolean connected;
     public ProtoChatWorker(IChatServer server, Socket connection) {
         this.server = server;
         this.connection = connection;
         try{
             output=connection.getOutputStream() ;//new ObjectOutputStream(connection.getOutputStream());
             input=connection.getInputStream(); //new ObjectInputStream(connection.getInputStream());
             connected=true;
         } catch (IOException e) {
             e.printStackTrace();
         }
     }

     public void run() {

         while(connected){
             try {
                // Object request=input.readObject();
                 System.out.println("Waiting requests ...");
                 ChatProtobufs.ChatRequest request=ChatProtobufs.ChatRequest.parseDelimitedFrom(input);
                 System.out.println("Request received: "+request);
                 ChatProtobufs.ChatResponse response=handleRequest(request);
                 if (response!=null){
                    sendResponse(response);
                 }
             } catch (IOException e) {
                 e.printStackTrace();
             }
             try {
                 Thread.sleep(1000);
             } catch (InterruptedException e) {
                 e.printStackTrace();
             }
         }
         try {
             input.close();
             output.close();
             connection.close();
         } catch (IOException e) {
             System.out.println("Error "+e);
         }
     }

     public void messageReceived(Message message) throws ChatException {
         System.out.println("Message received  "+message);
         try {
             sendResponse(ProtoUtils.createNewMessageResponse(message));
         } catch (IOException e) {
             throw new ChatException("Sending error: "+e);
         }
     }

     public void friendLoggedIn(User friend) throws ChatException {
         System.out.println("Friend logged in "+friend);
         try {
             sendResponse(ProtoUtils.createFriendLoggedInResponse(friend));
         } catch (IOException e) {
             e.printStackTrace();
         }
     }

     public void friendLoggedOut(User friend) throws ChatException {
         System.out.println("Friend logged out "+friend);
         try {
             sendResponse(ProtoUtils.createFriendLoggedOutResponse(friend));
         } catch (IOException e) {
             e.printStackTrace();
         }
     }

     private ChatProtobufs.ChatResponse handleRequest(ChatProtobufs.ChatRequest request){
         ChatProtobufs.ChatResponse response=null;
         switch (request.getType()){
             case Login:{
                 System.out.println("Login request ...");
                 User user=ProtoUtils.getUser(request);
                 try {
                     server.login(user, this);
                     return ProtoUtils.createOkResponse();
                 } catch (ChatException e) {
                     connected=false;
                     return ProtoUtils.createErrorResponse(e.getMessage());
                 }
             }
             case Logout:{
                 System.out.println("Logout request");
                 User user=ProtoUtils.getUser(request);
                 try {
                     server.logout(user, this);
                     connected=false;
                     return ProtoUtils.createOkResponse();

                 } catch (ChatException e) {
                     return ProtoUtils.createErrorResponse(e.getMessage());
                 }
             }
             case SendMessage:{
                 System.out.println("SendMessageRequest ...");
                 Message message=ProtoUtils.getMessage(request);
                 try {
                     server.sendMessage(message);
                     return ProtoUtils.createOkResponse();
                 } catch (ChatException e) {
                     return ProtoUtils.createErrorResponse(e.getMessage());
                 }
             }
             case GetLoggedFriends:{
                 System.out.println("GetLoggedFriends Request ...");
                 User user=ProtoUtils.getUser(request);
                 try {
                     User[] friends=server.getLoggedFriends(user);
                     return ProtoUtils.createLoggedFriendsResponse(friends);
                 } catch (ChatException e) {
                     return ProtoUtils.createErrorResponse(e.getMessage());
                 }
             }
         }
         return response;
     }

     private void sendResponse(ChatProtobufs.ChatResponse response) throws IOException{
         System.out.println("sending response "+response);
         response.writeDelimitedTo(output);
         //output.writeObject(response);
         output.flush();
     }
}
